main/src/ca/uwaterloo/flix/util/Validation.scala

```diff
@@ -169,14 +200,74 @@ object Validation {
     }
   }
 
+  def flatten[U, E](t1: Validation[Validation[U, E], E]): Validation[U, E] = t1 match {
+    case Success(Success(t)) =>
+      Success(t)
+    case Success(SuccessWithFailures(t, e)) =>
+      SuccessWithFailures(t, e)
+    case Success(Failure(e)) =>
+      Failure(e)
+    case SuccessWithFailures(Success(t), errors) =>
+      SuccessWithFailures(t, errors)
+    case SuccessWithFailures(SuccessWithFailures(t, e), errors) =>
+      SuccessWithFailures(t, errors #::: e)
+    case SuccessWithFailures(Failure(e), errors) =>
+      Failure(errors #::: e)
+    case Failure(errors) =>
+      Failure(errors)
+  }
+
+  def ap[T1, U, E](f: Validation[T1 => U, E])(t1: Validation[T1, E]): Validation[U, E] =
+    (f, t1) match {
+      case (Success(g), Success(v)) =>
+        Success(g(v))
+      case (Success(g), SuccessWithFailures(v, e2)) =>
+        SuccessWithFailures(g(v), e2)
+      case (Success(_), Failure(e2)) =>
+        Failure(e2)
+
+      case (SuccessWithFailures(g, e1), Success(v)) =>
+        SuccessWithFailures(g(v), e1)
+      case (SuccessWithFailures(g, e1), SuccessWithFailures(v, e2)) =>
+        SuccessWithFailures(g(v), e1 #::: e2)
+      case (SuccessWithFailures(_, e1), Failure(e2)) =>
+        Failure(e1 #::: e2)
+
+      case (Failure(e1), Success(_)) =>
+        Failure(e1)
+      case (Failure(e1), SuccessWithFailures(_, e2)) =>
+        Failure(e1 #::: e2)
+      case (Failure(e1), Failure(e2)) =>
+        Failure(e1 #::: e2)
+    }
+
+  private def curry[T1, T2, T3, T4, T5](f: (T1, T2, T3, T4) => T5): (T1, T2, T3) => T4 => T5 =
```

Comment:
Missing comments.

Reply:
Should we put these in a separate utility object? And maybe name them differently so we get nicer errors :)